### ISSUE TITLE
Multiprocessing

### ISSUE THREAD

**Issue by canada4663** [manahl/arctic#138]
I'm loving the library.  Great work - will contribute more as I use more!

As I'm loading my first large dataset, I noticed some opportunities for speed up at application level;  essentially my app is breaking large data frames into smaller frames by equity symbol, and writing the smaller frames to Arctic/Mongo - roughly 5k symbols - and is taking considerable time for historical option data.  As I expand to additional sources and the daily data loads become more time consuming, I would like to add multi-processing support so I can parallelize this operation to more fully saturate CPU and the underlying mongo-store.  Couple questions:

1) can you confirm there is not built in logic already inside Arctic to allow for multiprocessing / threading - seems that each write is blocking waiting for write confirmation from mongodb engine before returning (which makes sense to me!)

2) assuming I haven't missed this capability in Arctic as-is, my plan is to create a pool of Arctic instances that a process pool can pull from.  Does anyone have experience doing this - or potential other approaches recommended?

I will share any benchmark results / code snippets with the group if I can get it working and there is interest.

Ryan

**Comment by bmoscon** [comment]
1. There is not. The compression is done in parallel in the Cython code, but the writes are not done in parallel. 

2. No suggestions


There is benchmark support already (read benchmarks.md) so you can run those for single process and add new ones for multiprocess.

For the case you are describing, you might check out the new chunkstore  type. For your use case you'd only need to write a chunker that breaks the DF into the chunks you desire (by symbol) and it will handle the rest. It wouldnt be hard to add a multiprocessing component to chunkstore.

**Comment by richardbounds** [comment]
Arctic itself should be threadsafe, so it is no problem using a single Arctic instance across multiple threads to parallelise the i/o. If you want to run parallel processes, I'd give each worker process an Arctic instance and let them pull the symbols to load from a queue, or something like that. We do use Arctic a lot in multi-thread/process environments ourselves, so I'd expect no surprises here. 

I wouldn't try creating an Arctic instance, forking the process and then sharing the (forked) pymongo connection between multiple processes - I doubt that is guaranteed to work well (at the driver/socket level).

**Comment by bmoscon** [comment]
@richardbounds There would be an issue with multiprocessing if each process was writing to the same symbol, right? I don't see how multi-writing a symbol is thread safe. Even if only 1 does a write, and then everything else is an append, it still seems like there could be issues with updating the version document. Or am I missing something?

**Comment by richardbounds** [comment]
I'm assuming that we're talking about VersionStore here - TickStore is slightly different (but still threadsafe).

All data is versioned when it is written, and each version can be read back separately. We don't ever update versions of data that have been previously written (although old versions that have not been snapshotted are pruned out after a couple of hours to save space).

Because each version is immutable once written, we don't have to worry too much about collisions between multiple writers on the same symbol (each writer gets a unique version number for the data they wrote and can read it back with that version). The ordering of the versions is defined on the mongo primary when the version number is incremented, so for concurrent writes, last one wins.

Appends are only different in that they imply some ordering, which would need to be enforced by the application. Concurrent appends from different processes are treated the same way as writes, but because they append to the 'current' version from the perspective of the writing thread the final sequence of appends in the last version of the data could end up being scrambled.

In short, if you're appending to a symbol, don't append to the same symbol concurrently from multiple threads (or processes) without enforcing the ordering of the appends yourself (even then, probably not a good idea). If you're just writing (and reading the data back with the versions you wrote), concurrent write/read on the same symbol is OK, but can sometimes be confusing. Concurrent writes (or appends) to different symbols should always be fine.

**Comment by canada4663** [comment]
Gentlemen - thanks for the valuable inputs - I will make sure the application avoids concurrent writes / appends of the same symbol for sure... benchmarking it now in my application!

Ryan

**Comment by bmoscon** [comment]
@canada4663 is it ok to close this?

**Comment by canada4663** [comment]
yep!

I used thread pools which worked pretty well - got some speed up out to about 8 threads on a 2 CPU / 4 thread system.

Ryan

**Comment by bmoscon** [comment]
awesome!
